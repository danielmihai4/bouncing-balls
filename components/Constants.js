export const CANVAS_WIDTH = 800;
export const CANVAS_HEIGHT = 800;
export const MAX_SPEED = 10;
export const RADIUS = 5;
export const GRAVITY = 0.2;
export const DAMPING = 0.9;
export const TRACTION = 0.8;
export const BALL_COLOR = '#000000';

