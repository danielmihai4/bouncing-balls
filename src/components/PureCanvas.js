import uuid from 'uuid';
import React from 'react';
import { connect } from 'react-redux';
import { setMouseXCoordinate, setMouseYCoordinate, addBall } from '../actions/actions';
import { CANVAS_HEIGHT, CANVAS_WIDTH, MAX_SPEED } from './Constants';

export class PureCanvas extends React.Component {
    
    shouldComponentUpdate = () => {
      return false;
    }

    onAddBall = () => {
        this.props.addBall({
          id: uuid(),
          xCoordinate: this.props.mouseXCoordinate, 
          yCoordinate: this.props.mouseYCoordinate, 
          xVelocity: Math.floor(Math.random() * MAX_SPEED) + 1,
          yVelocity: - (Math.floor(Math.random() * MAX_SPEED) + 1)
        });
    }

    onMouseMove = (event) => {
      var bounds = event.target.getBoundingClientRect();
      var x = event.clientX - bounds.left;
      var y = event.clientY - bounds.top;

      this.props.setMouseXCoordinate(x);
      this.props.setMouseYCoordinate(y);
    };
  
    render() {
      return (
        <canvas
          width = {CANVAS_HEIGHT}
          height = {CANVAS_WIDTH}
          onClick = {this.onAddBall}
          onMouseMove = {this.onMouseMove}
          ref={node => node ? this.props.contextRef(node.getContext('2d')) : null}
        />
      );
    }
  }

const mapStateToProps = (state) => ({
    mouseXCoordinate: state.mouseCoordinates.xCoordinate, 
    mouseYCoordinate: state.mouseCoordinates.yCoordinate
});

const mapDispatchToProps = (dispatch) => ({
    setMouseXCoordinate: (xCoordinate) => dispatch(setMouseXCoordinate(xCoordinate)),
    setMouseYCoordinate: (yCoordinate) => dispatch(setMouseYCoordinate(yCoordinate)),
    addBall: (ball) => dispatch(addBall(ball))
});

export default connect(mapStateToProps, mapDispatchToProps)(PureCanvas);