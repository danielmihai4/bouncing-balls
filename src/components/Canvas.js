import React from 'react';
import { connect } from 'react-redux';
import PureCanvas from './PureCanvas';
import { BALL_COLOR, CANVAS_HEIGHT, CANVAS_WIDTH, RADIUS, GRAVITY, DAMPING, TRACTION } from './Constants';
import { removeBall } from '../actions/actions';

class Canvas extends React.Component {
  
  constructor(props) {
    super(props);
    this.saveContext = this.saveContext.bind(this);
  }

  saveContext = (ctx) => {
    this.ctx = ctx;
  }

  componentDidUpdate = () => {
    this.drawAxis();
    this.drawBalls();    
  }

  drawAxis = () => {
    const width = this.ctx.canvas.width;
    const height = this.ctx.canvas.height;

    this.ctx.save();
    this.ctx.beginPath();
    this.ctx.clearRect(0, 0, width, height);
    this.ctx.lineWidth = 5;
    this.ctx.moveTo(0, 0);
    this.ctx.lineTo(0, height);
    this.ctx.lineTo(width, height);
    this.ctx.stroke();    
    this.ctx.restore();
  }

  drawBalls = () => {
    if (this.props.balls.length === 0) {
      return;
    }

    this.ctx.moveTo(0,0);
    this.props.balls.forEach((ball) => {
      this.ctx.save();
      this.ctx.beginPath();
      this.ctx.arc(ball.xCoordinate, ball.yCoordinate, RADIUS, 0, 2 * Math.PI, false);      
      this.ctx.fillStyle = BALL_COLOR;
      this.ctx.fill();
      this.ctx.stroke();
      this.ctx.restore();      

      this.updateCoordinates(ball);      
    });
  }

  updateCoordinates = (ball) => {
    if (ball.xCoordinate + RADIUS >= CANVAS_WIDTH) {
      this.props.removeBall(ball.id);
      return;
    } else if (ball.xCoordinate - RADIUS <= 0) {
      ball.xVelocity = -ball.xVelocity * DAMPING;
      ball.xCoordinate = RADIUS;
    }

    if (ball.yCoordinate + RADIUS >= CANVAS_HEIGHT) {
      ball.yVelocity = -ball.yVelocity * DAMPING;
      ball.yCoordinate = CANVAS_HEIGHT - RADIUS;
      ball.xVelocity *= TRACTION;
    } else if (ball.yCoordinate - RADIUS <= 0) {
      ball.yVelocity = -ball.yVelocity * DAMPING;
      ball.yCoordinate = RADIUS;
    }
  
    ball.yVelocity += GRAVITY;
    ball.xCoordinate += ball.xVelocity;
    ball.yCoordinate += ball.yVelocity;
  }

  render() {
    return <PureCanvas contextRef={this.saveContext} />;
  }
}

const mapStateToProps = (state, props) => ({
  balls: state.balls 
});

const mapDispatchToProps = (dispatch, props) => ({
  removeBall: (id) => dispatch(removeBall(id))
});


export default connect(mapStateToProps, mapDispatchToProps)(Canvas);