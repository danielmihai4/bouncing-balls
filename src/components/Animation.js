import React from 'react';
import Canvas from './Canvas';

export class Animation extends React.Component {
    constructor(props) {
      super(props);
      this.updateAnimationState = this.updateAnimationState.bind(this);
    }
  
    componentDidMount = () => {
      this.rAF = requestAnimationFrame(this.updateAnimationState);
    }
  
    updateAnimationState = () => {
      this.rAF = requestAnimationFrame(this.updateAnimationState);
    }
  
    componentWillUnmount = () => {
      cancelAnimationFrame(this.rAF);
    }
  
    render() {
      return <Canvas />;
    }
  }

export default Animation;