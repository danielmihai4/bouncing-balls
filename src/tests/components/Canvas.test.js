import React from 'react';
import { shallow } from 'enzyme';
import { Canvas } from '../../components/Canvas';
import { CANVAS_WIDTH } from '../../components/Constants';

let wrapper, removeBall;

beforeEach(() => {
    removeBall = jest.fn();
    wrapper = shallow(<Canvas removeBall = {removeBall}/>);
});

test('should render Canvas correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should correctly update ball coordinates', () => {
    const ball = {
        id: '1',
        xCoordinate: 100,
        yCoordinate: 200,
        xVelocity: 1,
        yVelocity: -1
    };

    wrapper.instance().updateCoordinates(ball);

    expect(ball).toEqual({
        id: '1',
        xCoordinate: 101,
        yCoordinate: 199.2,
        xVelocity: 1,
        yVelocity: -0.8
    });
});

test('should correctly remove outside bounds ball', () => {
    const ball = {
        id: '1',
        xCoordinate: CANVAS_WIDTH + 1,
        yCoordinate: 200,
        xVelocity: 1,
        yVelocity: -1
    };

    wrapper.instance().updateCoordinates(ball);

    expect(removeBall).toHaveBeenLastCalledWith(ball.id);
});