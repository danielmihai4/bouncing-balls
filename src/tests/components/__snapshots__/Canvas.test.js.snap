// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Canvas correctly 1`] = `
<Connect(PureCanvas)
  contextRef={[Function]}
/>
`;
